### PR TITLE
Support grok2

### PR DESCRIPTION
**Pull request by JamesBrianD**
## Motivation
This PR introduces initial support for the Grok-2 model in SGLang-JAX. Fix SGL-33.

## Modifications

- Model compatibility updates: Adjustments to weight loading, layer definitions, and model configuration to accommodate Grok-2’s architecture.
- Tokenizer updates: Integration of a tiktoken-based tokenizer to support Grok-2 requirements.
- Code cleanup: Removal of redundant logs/prints, comment cleanup, formatting improvements, and general refactoring.

## Accuracy Tests
```
JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache uv run python -u -m sgl_jax.launch_server --model-path /models/xai-grok-2 \
--trust-remote-code --tp-size=32 --ep-size=8 --device=tpu --random-seed=3 --mem-fraction-static=0.8 --chunked-prefill-size=2048 \ 
--download-dir=/dev/shm --dtype=bfloat16 --max-running-requests=128 --skip-server-warmup \
--page-size=128 --tokenizer-path /models/xai-grok-2/tokenizer.tok.json  \
--dist-init-addr=$MASTER_IP:10011  --nnodes=8 --node-rank=$RANK
```
sky launch yaml
```
resources:
  accelerators: tpu-v6e-32
  accelerator_args:
      tpu_vm: True
      runtime_version: v2-alpha-tpuv6e
  use_spot: True # The default instance is spot.

file_mounts:
  ~/.ssh/id_rsa: ~/.ssh/id_rsa

setup: |
  chmod 600 ~/.ssh/id_rsa
  rm ~/.ssh/config
  git clone git@github.com:sgl-project/sglang-jax.git && cd sglang-jax  && git checkout support-grok2-rebase
  uv venv --python 3.12
  source .venv/bin/activate
  uv pip install -e "python[all]"
```
```
sky exec $(cat .cluster_name) -- "cd sglang-jax && source .venv/bin/activate && JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache uv run python -u -m sgl_jax.launch_server --model-path /models/xai-grok-2 --trust-remote-code --tp-size=32 --device=tpu --random-seed=3 --mem-fraction-static=0.8 --chunked-prefill-size=2048 --download-dir=/dev/shm --dtype=bfloat16 --max-running-requests=128 --skip-server-warmup --page-size=128 --tokenizer-path alvarobartt/grok-2-tokenizer --dist-init-addr=10.128.0.86:10011 --nnodes=8 --node-rank=\${SKYPILOT_NODE_RANK}"
```

- gsm8k result:
image

- test_flashattention kernel result:
image

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [ ] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/layers/layernorm.py [hunk 1]
```diff
@@ -212,3 +212,35 @@ def __call__(
...
+    mean2 = jnp.mean(lax.square(x_f32), axis=-1, keepdims=True)
+    y = jnp.asarray(x_f32 * lax.rsqrt(mean2 + epsilon), jnp.float32)
+    output = (y * jnp.asarray(weight, jnp.float32)).astype(orig_dtype)
+    if residual is None:
+        return output
+    else:
+        return output, residual
+
+
+def dual_rmsnorm_forward(
```
Comment (ispobock) [inline python/sgl_jax/srt/layers/layernorm.py:232]: Maybe we can fuse these ops into one kernel later.

Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/layernorm.py:232]: Ok, need to analyze it first to see if jax has already fused two rmsnorm calculations

**Comment by ispobock** [comment]
> Good idea, but one problem is how is the accuracy of the two-layer model measured. There is no way to test it using a test set.

Yes, we can just compare the logits after lm_head or sampled output_ids instead of test on the benchmark datasets. This is for limited CI resources consideration. Maybe we can add the whole model benchmark test in the nightly tests.

File: python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py [hunk 2]
```diff
@@ -143,6 +158,16 @@ def ref_ragged_paged_attention(
             mask = jnp.logical_or(mask, q_span - sliding_window >= kv_span)
         if soft_cap is not None:
             attn = soft_cap * jnp.tanh(attn / soft_cap)
+
+        # xai_temperature_scale: ref implementation from sgl-project/sglang
+        # python/sglang/srt/layers/attention/triton_ops/decode_attention.py
+        if xai_temperature_len is not None:
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py:164]: Add unit tests?

File: python/sgl_jax/srt/layers/activation.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import jax
+from flax import nnx
+
+
+class GeluAndMul(nnx.Module):
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/layers/activation.py:5]: Not used?

File: python/sgl_jax/srt/layers/logits_processor.py [hunk 4]
```diff
@@ -448,6 +448,7 @@ def _get_logits(
 
         logits = logits[:, : self.vocab_size] if logits.ndim > 1 else logits[: self.vocab_size]
 
-        if self.soft_cap:
+        if self.soft_cap and self.soft_cap > 0:
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/layers/logits_processor.py]: static_validate_inputs checks whether soft_cap is greater than 0.0, so this seems unnecessary here, and during assignment, 0 is recognized and set to None.

File: python/sgl_jax/srt/managers/tiktoken_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            messages=messages, add_generation_prompt=add_generation_prompt
+        )
+        return self.encode(ret) if tokenize else ret
+
+    def __call__(self, text: list[str], **kwargs):
+        return {
+            "input_ids": [self.encode(x) for x in text],
+        }
+
+    def init_xgrammar(self):
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/managers/tiktoken_tokenizer.py]: xgrammar isn’t supported yet—should we remove it for now?

File: python/sgl_jax/srt/models/grok.py [hunk 6]
```diff
@@ -0,0 +1,936 @@
...
+            )
+            inv_freq = jnp.array(
+                1.0 / (theta_scale_exponent ** (exponents / self.rotary_dim)),
+                dtype=jnp.float32,
+            )
+        else:
+            raise ValueError(f"Unknown extrapolation method: {self.extra_method}")
+        return inv_freq
+
+    def _compute_cos_sin_cache(self) -> jax.Array:
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/models/grok.py]: not used?

File: python/sgl_jax/srt/models/grok.py [hunk 7]
```diff
@@ -0,0 +1,936 @@
...
+                self.ffn = self.block_sparse_moe
+        else:
+            raise NotImplementedError()
+
+    def moe_with_rmoe(self, x):
+        """Combine MoE and residual MLP outputs (matches PyTorch implementation)."""
+        mlp_result = self.mlp(x)
+        moe_result = self.block_sparse_moe(x)
+        # Scale factor from the paper: 1/sqrt(2)
+        return (mlp_result + moe_result) / 1.4142135623730951
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/models/grok.py:545]: sqrt(2)？

File: python/sgl_jax/srt/models/grok.py [hunk 8]
```diff
@@ -0,0 +1,936 @@
...
+                    self.head_dim if not self.rope_rotate_half_dims else self.head_dim // 2
+                ),
+                max_position_embeddings=max_position,
+                base=int(self.rope_theta),
+                is_neox_style=True,
+                dtype=jnp.bfloat16,
+            )
+
+        logit_cap = max(getattr(config, "attn_logit_softcapping", 30.0), 0.0)
+        # logit_capping_method = getattr(config, "attn_logit_softcapping_method", "tanh")
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/models/grok.py]: not used?

File: python/sgl_jax/srt/layers/activation.py (hunk 3 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/activation.py:5]: Used in grok MLP.

File: python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py (hunk 2 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py:164]: Relevant unit tests will be added later

File: python/sgl_jax/srt/layers/logits_processor.py (hunk 4 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/logits_processor.py]: Get it.

File: python/sgl_jax/srt/models/grok.py (hunk 7 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/models/grok.py:545]: The official implementation of grok2 is this hard code

File: python/sgl_jax/srt/models/grok.py (hunk 6 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/models/grok.py]: Deleted.

File: python/sgl_jax/srt/models/grok.py (hunk 8 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/models/grok.py]: Deleted

File: python/sgl_jax/srt/managers/tiktoken_tokenizer.py (hunk 5 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/managers/tiktoken_tokenizer.py]: Deleted

File: python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py (hunk 2 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py:164]: Added

**Comment by JamesBrianD** [comment]
@ispobock @Kipsora @Bob-Chen222 @Prayer3th @agi-scaler Thanks a lot to for your contribution and help!
